### ISSUE TITLE
Does MobX work like Knockout's pureComputed? (Thinking of memory leaks)

### ISSUE THREAD

**Issue by danielearwicker** [mobxjs/mobx#711]
I used Knockout *very* heavily for a few years and really drove it to breaking point. It was fantastic but the only issue I hit was memory leaks which could be easy to introduce without realising it and then hard to track down. The cause was having some UI component bound to a model containing computeds, and then one day someone maintaining the code would make one of the computeds depend on some observable that was long-lived (e.g. the life time of the app!) This would make it enlist and so hang around for ever, and so would the UI component that was listening to it. Huge amounts of discarded objects (including heavyweight DOM elements) would build up quite quickly in a large, complex app.

In other words, merely reading certain values translated into "please keep me alive forever, oh and the same for everyone else who reads my value".

The solution came quite late in my usage of KO (mid-2014), but it was great:

http://knockoutjs.com/documentation/computed-pure.html

After that Knockout was near-perfect. Only the appearance of React took the shine off it, and I'm hoping MobX will serve the same purpose as KO's core features, in a React setting.

Reading some MobX blog posts, would I be correct in assuming that MobX is more like `pureComputed`, i.e. a computed value only listens to its dependencies when it itself being listened to?

**Comment by mweststrate** [comment]
that is correct :) `@computed` is only kept in sync and observers as long as it has observers on its own

**Comment by danielearwicker** [comment]
Awesome! A couple of years ago I ripped the innards out of Knockout and called it [Kor](https://github.com/danielearwicker/kor) (because it was the core of Knockout...) so I could use it with React, but then my day job switched me away from the web. Now I'm inching back to web I'm definitely going to give MobX some serious attention! It looks perfect.

**Comment by danielearwicker** [comment]
Hmm, further questions on this topic... The two modes of a `pureComputed` (what KO calls `asleep` and `listening`) can be switched back and forth any number of times. This is more handy than just being able to dispose, a one-way ticket. I assume the same in MobX `computed`?

This has the nice property that the same model can be wired into a UI component, which wakes it up (starts it listening) and when the UI component is detached the model goes back to sleep, but can wake up again later. It hasn't been permanently disposed - it never needs to be. While asleep, it can be abandoned and will be garbage collected.

In MobX there's `autorun` which is like a computed that stays listening, so it can run a side-effect whenever its dependencies notify, and it returns a one-way `disposer` function. This is like the old `ko.computed`.  The drawback is that it must be disposed at the appropriate time. So I guess this must be manually dealt with (e.g. `componentWillUnmount`) And disposal is one-way. It "ruins" the `autorun`, so I can't use it again.

Is there some way to unify these approaches? Like using a computed and then manually subscribing/unsubscribing during the period that it is wired into the UI? Is there an equivalent of the public `subscribe` method as found in KO? That would allow me to temporarily turn a `computed` into the equivalent of an `autorun`, for as long as I need to to be active, then switch it off until I need it again.

**Comment by mweststrate** [comment]
Yes, see observe 😊

Op di 20 dec. 2016 16:14 schreef Daniel Earwicker <notifications@github.com

**Comment by mweststrate** [comment]
@danielearwicker yes, computed works the same, it actively suspends and resumes, and even lazily evaluates if it is used while not being observed. So the effect on the UI is just as you described.

The approaches of autorun and computed are fundamentally not combinable; computeds produce _values_ while autoruns and friends produce _effects_. For values we can detect in a generic manner whether they are in use or not, so they can be suspended automatically. For effects there is no generic way to tell when the effect should stop occurring, so we need the user to dispose. Disposing is final indeed. Could technically probably be revived, but just creating a fresh autorun with the same function achieves the same.

Subscribing to changes can be done through `observe` (or  `autorun` / `reaction` etc of course).  Those all need disposing.

I hope that clarifies!

**Comment by danielearwicker** [comment]
Yep, perfect. Regarding:

> For effects there is no generic way to tell when the effect should stop occurring, so we need the user to dispose.

Yes, the model could be rebuilt each time, but when using KO it was a common source of bugs that something had been disposed in an attempt to "fix a memory leak" only to find that it was being reused later (this was in a pretty substantial complex app, 100,000+ lines with several maintainers). We switched to an approach where we never disposed anything, and as a result we had no memory leaks *and* no more premature-disposal problems.

So a `computed` that has been subscribed to (with `observe`), and evaluates to `undefined` because it has a side-effect, is that effectively, in practice, the same as `autorun`? If so, when I want to put it to sleep I just call the return value of `observe` (i.e. disposing this observation session, rather than the `computed` itself). Any reason why that wouldn't work?

**Comment by danielearwicker** [comment]
My testing indicates that it works as expected and judging from the API structure it clearly has to, i.e. instead of this:

```ts
const stop = autorun(() => console.log(`fullName is now ${person.fullName}`));
```

I could say:

```ts
const logger = computed(() => console.log(`fullName is now ${person.fullName}`));

let stop = logger.observe(ignore);
```

Either way, when `stop` is called it stops logging, but the second approach can "wake up" the same `computed`. OTOH as you say, if I encapsulate this properly then I should be able to use `autorun` and just recreate it each time.
